src/shogun/lib/type_case.h

```diff
@@ -0,0 +1,369 @@
+			ADD_TYPE_TO_MAP(float32_t , TYPE::PT_FLOAT32)
+			ADD_TYPE_TO_MAP(float64_t , TYPE::PT_FLOAT64)
+			ADD_TYPE_TO_MAP(floatmax_t , TYPE::PT_FLOATMAX)
+			ADD_TYPE_TO_MAP(complex128_t, TYPE::PT_COMPLEX128)
+			ADD_TYPE_TO_MAP(SGVector<float32_t>, TYPE::PT_SGVECTOR_FLOAT32)
+			ADD_TYPE_TO_MAP(SGVector<float64_t>, TYPE::PT_SGVECTOR_FLOAT64)
+			ADD_TYPE_TO_MAP(SGVector<floatmax_t>, TYPE::PT_SGVECTOR_FLOATMAX)
+			ADD_TYPE_TO_MAP(SGVector<int32_t>, TYPE::PT_SGVECTOR_INT32)
+			ADD_TYPE_TO_MAP(SGVector<int64_t>, TYPE::PT_SGVECTOR_INT64)
+			ADD_TYPE_TO_MAP(SGMatrix<float32_t>, TYPE::PT_SGMATRIX_FLOAT32)
+			ADD_TYPE_TO_MAP(SGMatrix<float64_t>, TYPE::PT_SGMATRIX_FLOAT64)
+			ADD_TYPE_TO_MAP(SGMatrix<floatmax_t>, TYPE::PT_SGMATRIX_FLOATMAX)
+			ADD_TYPE_TO_MAP(SGMatrix<int32_t>, TYPE::PT_SGMATRIX_INT32)
+			ADD_TYPE_TO_MAP(SGMatrix<int64_t>, TYPE::PT_SGMATRIX_INT64)
+	};
+	typemap sg_vector_types = {
+			ADD_TYPE_TO_MAP(SGVector<float32_t>, TYPE::PT_SGVECTOR_FLOAT32)
+			ADD_TYPE_TO_MAP(SGVector<float64_t>, TYPE::PT_SGVECTOR_FLOAT64)
+			ADD_TYPE_TO_MAP(SGVector<floatmax_t>, TYPE::PT_SGVECTOR_FLOATMAX)
+			ADD_TYPE_TO_MAP(SGVector<int32_t>, TYPE::PT_SGVECTOR_INT32)
+			ADD_TYPE_TO_MAP(SGVector<int64_t>, TYPE::PT_SGVECTOR_INT64)
+	};
+	typemap sg_matrix_types = {
+			ADD_TYPE_TO_MAP(SGMatrix<float32_t>, TYPE::PT_SGMATRIX_FLOAT32)
+			ADD_TYPE_TO_MAP(SGMatrix<float64_t>, TYPE::PT_SGMATRIX_FLOAT64)
+			ADD_TYPE_TO_MAP(SGMatrix<floatmax_t>, TYPE::PT_SGMATRIX_FLOATMAX)
+			ADD_TYPE_TO_MAP(SGMatrix<int32_t>, TYPE::PT_SGMATRIX_INT32)
+			ADD_TYPE_TO_MAP(SGMatrix<int64_t>, TYPE::PT_SGMATRIX_INT64)
+	};
+	typemap sg_non_complex_types = {
+			ADD_TYPE_TO_MAP(bool, TYPE::PT_BOOL)
+			ADD_TYPE_TO_MAP(char, TYPE::PT_CHAR)
+			ADD_TYPE_TO_MAP(int8_t, TYPE::PT_INT8)
+			ADD_TYPE_TO_MAP(uint8_t, TYPE::PT_UINT8)
+			ADD_TYPE_TO_MAP(int16_t , TYPE::PT_INT16)
+			ADD_TYPE_TO_MAP(uint16_t , TYPE::PT_UINT16)
+			ADD_TYPE_TO_MAP(int32_t , TYPE::PT_INT32)
+			ADD_TYPE_TO_MAP(uint32_t , TYPE::PT_UINT32)
+			ADD_TYPE_TO_MAP(int64_t , TYPE::PT_INT64)
+			ADD_TYPE_TO_MAP(uint64_t , TYPE::PT_UINT64)
+			ADD_TYPE_TO_MAP(float32_t , TYPE::PT_FLOAT32)
+			ADD_TYPE_TO_MAP(float64_t , TYPE::PT_FLOAT64)
+			ADD_TYPE_TO_MAP(floatmax_t , TYPE::PT_FLOATMAX)
+	};
+	typemap sg_real_types = {
+			ADD_TYPE_TO_MAP(float32_t , TYPE::PT_FLOAT32)
+			ADD_TYPE_TO_MAP(float64_t , TYPE::PT_FLOAT64)
+			ADD_TYPE_TO_MAP(floatmax_t , TYPE::PT_FLOATMAX)
+	};
+	typemap sg_non_integer_types = {
+			ADD_TYPE_TO_MAP(float32_t , TYPE::PT_FLOAT32)
+			ADD_TYPE_TO_MAP(float64_t , TYPE::PT_FLOAT64)
+			ADD_TYPE_TO_MAP(floatmax_t , TYPE::PT_FLOATMAX)
+			ADD_TYPE_TO_MAP(complex128_t, TYPE::PT_COMPLEX128)
+	};
+#undef ADD_TYPE_TO_MAP
+
+	/** Checks the underlying type of an Any instance and
+	 * executes the appropriate lambda. The type is checked
+	 * against a typemap and if it isn't found there
+	 * a ShogunException is thrown.
+	 * The lambda function must have a specific signature,
```

Comment:
I don't think so. Because the compiler knows what type is required by the function inside the lambda, i.e. summaryValue->set_simple_value() should be a float, but at compile time we don't know what the dispatched object is, so we compile functions for each possibility, including vector and matrix. That's why knowing if it's scalar, vector or matrix helps out a lot with SFINAE and avoid a compile error.

Reply:
The only way would be to check the signature of all function calls inside the lambda and use that to figure out if it's scalar, vector or matrix, but I don't know if that is possible!

Reply:
It is clear that the caller needs to know what he wants to dispatch to (primitive,vector/matrix).
BUT, what if the call is the same for all, and then the compiler uses the signature of the lambda to decide. This means we would have to change the signature of course.

* define lambda (with SGVector in signature somehow)
* pass to `sg_any_dispatch`
* compiler sees lambda and knows that it is a vector and does the appropriate call

If that is not possible, I think 3 different functions is the best compromise. Or we ask @lisitsyn if he has an idea (though he is busy at nips until next week)

Reply:
That should be doable! It's a matter of API design, functionally that is almost the same as we have now.

Reply:
I added some utility structs to figure out if the type is SGVector or SGMatrix, so can use that to execute the correct lambda